core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -241,6 +246,58 @@ static MetaData updateMetaDataWithRoutingTable(MetaData oldMetaData, RoutingTabl
     }
 
     /**
+     * Possibly removes delay markers from shards and reroutes if markers have been deleted.
+     * <p>
+     * If the same instance of the routing table is returned, then no change has been made.</p>
+     */
+    public RoutingAllocation.Result applyDelayedShards(ClusterState clusterState) {
+        return applyDelayedShards(clusterState, true);
+    }
+
+    public RoutingAllocation.Result applyDelayedShards(ClusterState clusterState, boolean withReroute) {
+        RoutingNodes routingNodes = getMutableRoutingNodes(clusterState);
+        RoutingAllocation allocation = new RoutingAllocation(allocationDeciders, routingNodes, clusterState.nodes(), clusterInfoService.getClusterInfo(), currentNanoTime());
+        boolean changed = removeDelayMarkers(allocation);
```

Comment:
This makes delayed scheduling efficient (same approach as for Shard started / shard failed). If we haven't removed a delay marker, we don't need to go through reroute and do all the shard balancing...